example-github-with-coroutine/src/main/java/example/github/GitHubExample.kt

```diff
@@ -0,0 +1,133 @@
+import feign.codec.ErrorDecoder
+import feign.gson.GsonEncoder
+import feign.kotlin.CoroutineFeign
+import java.io.IOException
+import java.util.concurrent.TimeUnit
+
+suspend fun main() {
+    val github = GitHub.connect()
+    println("Let's fetch and print a list of the contributors to this org.")
+    val contributors = github.contributors("openfeign")
+    for (contributor in contributors) {
+        println(contributor)
+    }
+    println("Now, let's cause an error.")
+    try {
+        github.contributors("openfeign", "some-unknown-project")
+    } catch (e: GitHubClientError) {
+        println(e.message)
+    }
+    println("Now, try to create an issue - which will also cause an error.")
+    try {
+        val issue = GitHub.Issue(
+            title = "The title",
+            body = "Some Text",
+        )
+        github.createIssue(issue, "OpenFeign", "SomeRepo")
+    } catch (e: GitHubClientError) {
+        println(e.message)
+    }
+}
+
+/**
+ * Inspired by `com.example.retrofit.GitHubClient`
+ */
+
+interface GitHub {
+    data class Repository(
+        val name: String
+    )
+
+    data class Contributor(
+        val login: String
+    )
+
+    data class Issue(
+        val title: String,
+        val body: String,
+        val assignees: List<String> = emptyList(),
+        val milestone: Int = 0,
+        val labels: List<String> = emptyList(),
+    )
+
+    @RequestLine("GET /users/{username}/repos?sort=full_name")
+    suspend fun repos(@Param("username") owner: String): List<Repository>
+
+    @RequestLine("GET /repos/{owner}/{repo}/contributors")
+    suspend fun contributors(@Param("owner") owner: String, @Param("repo") repo: String): List<Contributor>
+
+    @RequestLine("POST /repos/{owner}/{repo}/issues")
+    suspend fun createIssue(issue: Issue, @Param("owner") owner: String, @Param("repo") repo: String)
+
+    companion object {
+        fun connect(): GitHub {
+            val decoder: Decoder = feign.gson.GsonDecoder()
+            val encoder: Encoder = GsonEncoder()
+            return CoroutineFeign.coBuilder<Unit>()
+                .encoder(encoder)
+                .decoder(decoder)
+                .errorDecoder(GitHubErrorDecoder(decoder))
+                .logger(ErrorLogger())
+                .logLevel(Logger.Level.BASIC)
+                .requestInterceptor { template ->
+                    template.header(
+                        // not available when building PRs...
+                        // https://docs.travis-ci.com/user/environment-variables/#defining-encrypted-variables-in-travisyml
+                        "Authorization",
+                        "token 383f1c1b474d8f05a21e7964976ab0d403fee071");
+                }
+                .options(Request.Options(10, TimeUnit.SECONDS, 60, TimeUnit.SECONDS, true))
+                .target(GitHub::class.java, "https://api.github.com")
+        }
+    }
+}
+
+/** Lists all contributors for all repos owned by a user.  */
```

Comment:
Currently, when this extension function is declared inside the GitHub interface, a Contracts parsing error occurs.

```
Exception in thread "main" java.lang.IllegalStateException: Method GitHub#contributors(String,Continuation) not annotated with HTTP method type (ex. GET, POST)
Warnings:
- Class GitHub has annotations [Metadata] that are not used by contract Default
	at feign.Util.checkState(Util.java:136)
	at feign.Contract$BaseContract.parseAndValidateMetadata(Contract.java:115)
	at feign.Contract$BaseContract.parseAndValidateMetadata(Contract.java:65)
	at feign.DeclarativeContract.parseAndValidateMetadata(DeclarativeContract.java:38)
```


This is because it is recognized as a non-default method, so contracts are required. (The default method of the kotlin interface is compiled as an abstract method of the java interface.)

So, I should use `ReflectJvmMapping.getKotlinFunction(method).isAbstract() == false` to check if it is the default method.

However, it seems that many code modifications may be required to put this logic in the code below without using optional dependency.

https://github.com/OpenFeign/feign/blob/75a3c1cf6f0e57c2d1eaa5f8be2538489e605650/core/src/main/java/feign/Util.java#L143-L153

Reply:
Yeah, forget the optional dependency, that ain't optional and ain't a option for feign core =)

The feign approach for this kind of problems is abstraction.  So would need an interface for `DefaultMethodsDetection/Handling/Operator`, a default implementation as above and a kotlin impl that does what kotlin needs

Reply:
I totally agree with the approach you are talking about. Let's consider improving it with a separate PR.